core/slice/heap/heap.odin

```diff
@@ -0,0 +1,231 @@
+/*
+	Copyright 2022 Dale Weiler <weilercdale@gmail.com>.
+	Made available under Odin's BSD-3 license.
+
+	List of contributors:
+		Dale Weiler: Initial implementation
+*/
+
+// Package implements a generic max heap in-place on a slice for any type.
+package heap
+
+/*
+	Constructs a max heap in slice given by data with comparator. A max heap is
+	a range of elements which has the following properties:
+
+	1. With N = len(data), for all 0 < i < N, data[(i - 1) / 2] does not compare
+	   less than data[i].
+
+	2. A new element can be added using push in O(log n) time.
+
+	3. The first element can be removed using pop in O(log n) time.
+
+	The comparator compares elements of type T and can be used to construct a
+	max heap (less than) or min heap (greater than) for T.
+*/
+make :: proc(data: []$T, compare: $C) {
```

Comment:
Why is this called `make`? And not something like `construct` or `init` or whatever?

Reply:
Because all the other implementations call it make and I think it's called that in the algorithm? It's not a one time construction of something, you're expected to use it multiple times on the same array to make it into a heap because the heap property is easily invalidated under mutation so it's often used multiple time when you need the heap property.

Reply:
The reason `make` feels off is because we have `make` in Odin and it has a specific meaning. So this having `make` and not returning anything is a bit off.

Reply:
Other implementations call it heapify, but `heap.heapify` seems silly to me.